kubeconfig/kubeconfig.go

```diff
@@ -0,0 +1,175 @@
+	host, port := os.Getenv("KUBERNETES_SERVICE_HOST"), os.Getenv("KUBERNETES_SERVICE_PORT")
+	if host == "" || port == "" {
+		return ""
+	}
+
+	server := "https://" + net.JoinHostPort(host, port)
+	kubeconfig := generateKubeConfig(token, server, namespace)
+
+	bytes, err := yaml.Marshal(&kubeconfig)
+	if err != nil {
+		logrus.Error("error: %v", err)
+		return ""
+	}
+	return string(bytes)
+}
+
+// CreateKubeConfig creates a kubeconfig located at $KUBECONFIG if set.
+// If it is not set then fall back to $HOME/.kube/config
+func CreateKubeConfig(cmdRslv exec_info.InfoExecCreator, namespace string, token string, containerInfo *model.ContainerInfo) error {
+	config := createKubeConfigText(token, namespace)
+
+	kubeconfigDir, err := findKubeConfigDir(cmdRslv, containerInfo)
+	if err != nil {
+		return err
+	}
+
+	err = createKubeConfigDir(cmdRslv, kubeconfigDir, containerInfo)
+	if err != nil {
+		return err
+	}
+
+	err = syncKubeConfig(cmdRslv, config, kubeconfigDir, containerInfo)
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+func findKubeConfigDir(cmdRslv exec_info.InfoExecCreator, containerInfo *model.ContainerInfo) (string, error) {
+	infoExec := cmdRslv.CreateInfoExec([]string{"sh", "-c", "echo $KUBECONFIG"}, containerInfo)
+	if err := infoExec.Start(); err != nil {
+		logrus.Debugf("Error is not available in %s/%s. Error: %s", containerInfo.PodName, containerInfo.ContainerName, err.Error())
+		return "", errors.New("Could not retrieve $KUBECONFIG")
+	}
+	kubeconfigDir := infoExec.GetOutput()
+	if kubeconfigDir != "\n" {
+		return strings.Replace(kubeconfigDir, "/config", "", 1), nil
+	}
+
+	infoExec = cmdRslv.CreateInfoExec([]string{"sh", "-c", "echo $HOME"}, containerInfo)
+	if err := infoExec.Start(); err != nil {
+		logrus.Debugf("Error is not available in %s/%s. Error: %s", containerInfo.PodName, containerInfo.ContainerName, err.Error())
+		return "", errors.New("Could not retrieve $HOME")
+	}
+	kubeconfigDir = strings.TrimSuffix(infoExec.GetOutput(), "\n")
+	return kubeconfigDir + "/.kube", nil
+}
+
+func createKubeConfigDir(cmdRslv exec_info.InfoExecCreator, kubeconfigDir string, containerInfo *model.ContainerInfo) error {
+	infoExec := cmdRslv.CreateInfoExec([]string{"sh", "-c", "mkdir -p " + kubeconfigDir}, containerInfo)
+	if err := infoExec.Start(); err != nil {
+		logrus.Debugf("Error is not available in %s/%s. Error: %s", containerInfo.PodName, containerInfo.ContainerName, err.Error())
+		return errors.New("Could not create directory: " + kubeconfigDir)
+	}
+	return nil
+}
+
+func syncKubeConfig(cmdRslv exec_info.InfoExecCreator, config string, kubeconfigLocation string, containerInfo *model.ContainerInfo) error {
+	infoExec := cmdRslv.CreateInfoExec([]string{"sh", "-c", "echo \"" + config + "\" > " + kubeconfigLocation + "/config"}, containerInfo)
+	if err := infoExec.Start(); err != nil {
+		logrus.Debugf("Error is not available in %s/%s. Error: %s", containerInfo.PodName, containerInfo.ContainerName, err.Error())
```

Comment:
`Error is not available`?... Should it be something like - failed to write kubeconfig file $kubeconfigLocation ...?